safir/tests/event_manager_test.py

```diff
@@ -0,0 +1,300 @@
+"""Test metrics tooling."""
+
+import asyncio
+import math
+from uuid import UUID
+
+import pytest
+from aiokafka import AIOKafkaConsumer
+from aiokafka.admin.client import AIOKafkaAdminClient, NewTopic
+from faststream.kafka import KafkaBroker
+from pydantic import AnyUrl, Field
+from schema_registry.client.client import AsyncSchemaRegistryClient
+from schema_registry.serializers.message_serializer import (
+    AsyncAvroMessageSerializer,
+)
+
+from safir.kafka import (
+    KafkaConnectionSettings,
+    PydanticSchemaManager,
+    SchemaManagerSettings,
+    SecurityProtocol,
+)
+from safir.metrics import (
+    DuplicateEventError,
+    EventDependency,
+    EventMaker,
+    EventManager,
+    EventManagerUnintializedError,
+    EventMetadata,
+    EventPayload,
+    EventPublisher,
+    KafkaTopicError,
+    MetricsConfiguration,
+    MetricsConfigurationWithKafka,
+)
+
+
+class MyEvent(EventPayload):
+    """An event payload."""
+
+    foo: str
+
+
+class Events(EventMaker):
+    """A class to hold event publishers."""
+
+    async def initialize(self, manager: EventManager) -> None:
+        self.my_event = await manager.create_publisher("myevent", MyEvent)
+
+
+async def subscribe_and_wait(consumer: AIOKafkaConsumer, pattern: str) -> None:
+    """Avoid race condition when subscribing to topics.
+
+    There is some race condition where an exception is raised if we start
+    trying to consume messages too soon after we have subscribed to a topic
+    or pattern, because partitions haven't been assigned to the consumer yet.
+    """
+    consumer.subscribe(pattern=pattern)
+    await asyncio.sleep(0.5)
+
+
+async def assert_from_kafka(
+    consumer: AIOKafkaConsumer,
+    schema_registry: AsyncSchemaRegistryClient,
+    event: EventPublisher[MyEvent],
+    foo: str,
+) -> None:
+    """Grab a message from kafka, deserialize it, and assert stuff about it."""
+    message = await consumer.getone()
+    assert isinstance(message.value, bytes)
+    serializer = AsyncAvroMessageSerializer(schema_registry)
+    deserialized_dict = await serializer.decode_message(message.value)
+    assert deserialized_dict is not None
+    deserialized = event.event_class(**deserialized_dict)
+
+    assert isinstance(deserialized, EventMetadata)
+    assert isinstance(deserialized.id, UUID)
+    assert deserialized.app_name == "testapp"
+
+    # dataclasses-avroschema serializes python ``datetime``s into avro
+    # ``timestamp-millis``s
+    """Publish events to actual storage and read them back and verify them."""
+    assert (
+        math.trunc(deserialized.timestamp_ns / 1e6)
+        == deserialized.timestamp.timestamp() * 1e3
+    )
```

Comment:
This is super, super minor but I find this formatting of assert really annoying, so I use a temporary variable to avoid it.
```suggestion
    expected = deserialized.timestamp.timestamp() * 1e3
    assert math.trunc(deserialized.timestamp_ns / 1e6) == expected
```